presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RewriteCaseExpressionPredicate.java

```diff
@@ -0,0 +1,223 @@
+import com.facebook.presto.spi.relation.SpecialFormExpression;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.relational.Expressions;
+import com.facebook.presto.sql.relational.FunctionResolution;
+import com.facebook.presto.sql.relational.RowExpressionDeterminismEvaluator;
+import com.facebook.presto.sql.tree.ComparisonExpression;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.SystemSessionProperties.isOptimizeCaseExpressionPredicate;
+import static com.facebook.presto.common.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.SWITCH;
+import static com.facebook.presto.spi.relation.SpecialFormExpression.Form.WHEN;
+import static com.facebook.presto.sql.planner.plan.Patterns.filter;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Re-writes CASE expression predicate into a series of OR clauses
+ * <p>
+ * (CASE
+ * WHEN condition1 THEN result1
+ * WHEN condition2 THEN result2
+ * ELSE elseResult
+ * END) = value
+ * <p>
+ * becomes
+ * <p>
+ * (result1 = value AND condition1) OR
+ * (result2 = value AND condition2) OR
+ * (elseResult = value AND (!condition1 AND !condition2))
+ * <p>
+ * This conversion is done so that it is easy for the ExpressionInterpreter & other Optimizers to further
+ * simplify this.
+ * i.e, ExpressionInterpreter can discard all conditions in which result != value
+ */
+public class RewriteCaseExpressionPredicate
+        implements Rule<FilterNode>
+{
+    private static final Pattern<FilterNode> PATTERN = filter();
+    private final FunctionAndTypeManager functionAndTypeManager;
+
+    public RewriteCaseExpressionPredicate(FunctionAndTypeManager functionAndTypeManager)
+    {
+        this.functionAndTypeManager = requireNonNull(functionAndTypeManager, "functionManager is null");
+    }
+
+    private static class Rewriter
+            extends RowExpressionRewriter<Void>
+    {
+        private final FunctionAndTypeManager functionAndTypeManager;
+        private final FunctionResolution functionResolution;
+        private final LogicalRowExpressions logicalRowExpressions;
+
+        private Rewriter(FunctionAndTypeManager functionAndTypeManager)
+        {
+            this.functionAndTypeManager = functionAndTypeManager;
+            this.functionResolution = new FunctionResolution(functionAndTypeManager);
+            this.logicalRowExpressions = new LogicalRowExpressions(new RowExpressionDeterminismEvaluator(this.functionAndTypeManager),
+                    this.functionResolution,
+                    this.functionAndTypeManager);
+        }
+
+        @Override
+        public RowExpression rewriteCall(CallExpression node, Void context, RowExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (functionResolution.isComparisonFunction(node.getFunctionHandle())) {
+                if (node.getArguments().size() == 2) {
```

Comment:
Combine the two if's into AND for better readability

Reply:
Done.